File: ruby/example_code/elastictranscoder/SqsQueueNotificationWorker.rb
# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: Apache-2.0

# Purpose:
# JobStatusNotificationSample.rb demonstrates how to create a queue for handling notifications
# for an Amazon Elastic Transcoder job using the AWS SDK for Ruby.

# snippet-start:[elastictranscoder.ruby.create_sqs_notification_queue.import]
require 'aws-sdk-elastictranscoder'
require 'aws-sdk-sqs'

# Class which will poll SQS for job state notifications in a separate thread.
# This class is intended for batch-processing.  If you are using a ruby-on-rails
# server, then a better implementation would be to subscribe your server
# directly to the notification topic and process the SNS messages directly.
class SqsQueueNotificationWorker
  def initialize(region, queue_url, options = {})
    options = { max_messages: 5, visibility_timeout: 15, wait_time_seconds: 5 }.merge(options)
    @queue_url = queue_url
    @max_messages = options[:max_messages]
    @visibility_timeout = options[:visibility_timeout]
    @wait_time_seconds = options[:wait_time_seconds]
    @handlers = []
  end

  def start
    @shutdown = false
    @thread = Thread.new { poll_and_handle_messages }
  end

  def stop
    @shutdown = true
  end

  def add_handler(handler)
    @handlers << handler
  end

  def remove_handler(handler)
    @handlers -= [handler]
  end

  def poll_and_handle_messages
    sqs_client = Aws::SQS::Client.new
    until @shutdown
      sqs_messages = sqs_client.receive_message(
        queue_url: @queue_url,
        max_number_of_messages: @max_messages,
        wait_time_seconds: @wait_time_seconds
      ).data[:messages]

      next if sqs_messages.nil?

      sqs_messages.each do |sqs_message|
        notification = JSON.parse(JSON.parse(sqs_message[:body])['Message'])
        @handlers.each do |handler|
          handler.call(notification)
          sqs_client.delete_message(queue_url: @queue_url, receipt_handle: sqs_message[:receipt_handle])
        end
      end
    end
  end
end
# snippet-end:[elastictranscoder.ruby.create_sqs_notification_queue.import]

File: java/example_code/elastictranscoder/src/main/java/aws/example/elastictranscoder/utils/SqsQueueNotificationWorker.java
//snippet-sourcedescription:[SqsQueueNotificationWorker.java provides a worker for handling job notifications.]
//snippet-keyword:[Java]
//snippet-keyword:[Code Sample]
//snippet-keyword:[Amazon Elastic Transcoder]
//snippet-service:[elastictranscoder]
//snippet-sourcetype:[full-example]
//snippet-sourcedate:[]
//snippet-sourceauthor:[AWS]
/*
 * Copyright 2010-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 *  http://aws.amazon.com/apache2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */
// snippet-start:[elastictranscoder.java.sqs_queue_notification_worker.import]
package com.amazonaws.services.elastictranscoder.samples.utils;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import com.amazonaws.services.elastictranscoder.samples.model.JobStatusNotification;
import com.amazonaws.services.elastictranscoder.samples.model.JobStatusNotificationHandler;
import com.amazonaws.services.elastictranscoder.samples.model.Notification;
import com.amazonaws.services.sqs.AmazonSQS;
import com.amazonaws.services.sqs.model.DeleteMessageRequest;
import com.amazonaws.services.sqs.model.Message;
import com.amazonaws.services.sqs.model.ReceiveMessageRequest;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;

/**
 * Polls an SQS queue for Elastic Transcoder job status notification messages
 * and calls the handle method on all registered JobStatusNotificationHandler
 * objects before deleting the message from the queue.
 */
public class SqsQueueNotificationWorker implements Runnable {

    private static final int MAX_NUMBER_OF_MESSAGES = 5;
    private static final int VISIBILITY_TIMEOUT = 15;
    private static final int WAIT_TIME_SECONDS = 15;
    private static final ObjectMapper mapper = new ObjectMapper();
    
    private AmazonSQS amazonSqs;
    private String queueUrl;
    private List<JobStatusNotificationHandler> handlers;
    
    private volatile boolean shutdown = false;
    
    public SqsQueueNotificationWorker(AmazonSQS amazonSqs, String queueUrl) {
        this.amazonSqs = amazonSqs;
        this.queueUrl = queueUrl;
        this.handlers = new ArrayList<JobStatusNotificationHandler>();
    }
    
    public void addHandler(JobStatusNotificationHandler jobStatusNotificationHandler) {
        synchronized(handlers) {
            this.handlers.add(jobStatusNotificationHandler);
        }
    }
    
    public void removeHandler(JobStatusNotificationHandler jobStatusNotificationHandler) {
        synchronized(handlers) {
            this.handlers.remove(jobStatusNotificationHandler);
        }
    }
    
    @Override
    public void run() {
        ReceiveMessageRequest receiveMessageRequest = new ReceiveMessageRequest()
            .withQueueUrl(queueUrl)
            .withMaxNumberOfMessages(MAX_NUMBER_OF_MESSAGES)
            .withVisibilityTimeout(VISIBILITY_TIMEOUT)
            .withWaitTimeSeconds(WAIT_TIME_SECONDS);
        
        while (!shutdown) {
            // Long pole the SQS queue.  This will return as soon as a message
            // is received, or when WAIT_TIME_SECONDS has elapsed.
            List<Message> messages = amazonSqs.receiveMessage(receiveMessageRequest).getMessages();
            if (messages == null) {
                // If there were no messages during this poll period, SQS
                // will return this list as null.  Continue polling.
                continue;
            }
            
            synchronized(handlers) {
                for (Message message : messages) {
                    try {
                        // Parse notification and call handlers.
                        JobStatusNotification notification = parseNotification(message);
                        for (JobStatusNotificationHandler handler : handlers) {
                            handler.handle(notification);
                        }
                    } catch (IOException e) {
                        System.out.println("Failed to convert notification: " + e.getMessage());
                    }
                    
                    // Delete the message from the queue.
                    amazonSqs.deleteMessage(new DeleteMessageRequest().withQueueUrl(queueUrl).withReceiptHandle(message.getReceiptHandle()));
                }
            }
        }
    }
    
    private JobStatusNotification parseNotification(Message message) throws IOException {
        Notification<JobStatusNotification> notification = mapper.readValue(message.getBody(), new TypeReference<Notification<JobStatusNotification>>() {});
        return notification.getMessage();
    }

    public void shutdown() {
        shutdown = true;
    }
}
// snippet-end:[elastictranscoder.java.sqs_queue_notification_worker.import]
